xarray/core/duck_array_ops.py

```diff
@@ -221,9 +222,17 @@ def asarray(data, xp=np):
 
 def as_shared_dtype(scalars_or_arrays, xp=np):
     """Cast a arrays to a shared dtype using xarray's type promotion rules."""
-    array_type_cupy = array_type("cupy")
-    if array_type_cupy and any(
-        isinstance(x, array_type_cupy) for x in scalars_or_arrays
+    if any(is_extension_array_dtype(x) for x in scalars_or_arrays):
+        extension_array_types = [
+            x.dtype for x in scalars_or_arrays if is_extension_array_dtype(x)
+        ]
+        if len(extension_array_types) == len(scalars_or_arrays) and all(
+            isinstance(x, type(extension_array_types[0])) for x in extension_array_types
+        ):
+            return scalars_or_arrays
```

Comment:
Are we missing raising an error here if there is a combo of extension array and numpy arrays?

Reply:
I don't know.  I guess we probably should?  It would be easier to undo that than to stop allowing it for some reason.

Reply:
I think one thing this allows us to do is mix and match categoricals with np array object dtypes.  This could be desireable in some scenarios...

Reply:
I will remove.